aptos-move/move-examples/data_structures/sources/smart_table.move

```diff
@@ -0,0 +1,354 @@
+        // partition the bucket. after the loop, i == j and [0..i) stays in old bucket, [j..len) goes to new bucket
+        let i = 0;
+        let j = vector::length(old_bucket);
+        let len = j;
+        while (i < j) {
+            let entry = vector::borrow(old_bucket, i);
+            let index = bucket_index(table.level, table.num_buckets, entry.hash);
+            if (index == new_bucket_index) {
+                j = j - 1;
+                vector::swap(old_bucket, i, j);
+            } else {
+                i = i + 1;
+            };
+        };
+        while (j < len) {
+            let entry = vector::pop_back(old_bucket);
+            vector::push_back(&mut new_bucket, entry);
+            len = len - 1;
+        };
+        table_with_length::add(&mut table.buckets, new_bucket_index, new_bucket);
+    }
+
+    /// Return the expected bucket index to find the hash.
+    /// Basically, it use different base `1 << level` vs `1 << (level + 1)` in modulo operation based on the target
+    /// bucket index compared to the index of the next bucket to split.
+    fun bucket_index(level: u8, num_buckets: u64, hash: u64): u64 {
+        let index = hash % (1 << (level + 1));
+        if (index < num_buckets) {
+            // in existing bucket
+            index
+        } else {
+            // in unsplitted bucket
+            index % (1 << level)
+        }
+    }
+
+    /// Acquire an immutable reference to the value which `key` maps to.
+    /// Aborts if there is no entry for `key`.
+    public fun borrow<K: drop, V>(table: &SmartTable<K, V>, key: K): &V {
+        let index = bucket_index(table.level, table.num_buckets, sip_hash_from_value(&key));
+        let bucket = table_with_length::borrow(&table.buckets, index);
+        let i = 0;
+        let len = vector::length(bucket);
+        while (i < len) {
+            let entry = vector::borrow(bucket, i);
+            if (&entry.key == &key) {
+                return &entry.value
+            };
+            i = i + 1;
+        };
+        abort error::invalid_argument(ENOT_FOUND)
+    }
+
+    /// Acquire a mutable reference to the value which `key` maps to.
+    /// Aborts if there is no entry for `key`.
+    public fun borrow_mut<K: drop, V>(table: &mut SmartTable<K, V>, key: K): &mut V {
+        let index = bucket_index(table.level, table.num_buckets, sip_hash_from_value(&key));
+        let bucket = table_with_length::borrow_mut(&mut table.buckets, index);
+        let i = 0;
+        let len = vector::length(bucket);
+        while (i < len) {
+            let entry = vector::borrow_mut(bucket, i);
+            if (&entry.key == &key) {
+                return &mut entry.value
+            };
+            i = i + 1;
+        };
+        abort error::invalid_argument(ENOT_FOUND)
+    }
+
+    /// Returns true iff `table` contains an entry for `key`.
+    public fun contains<K: drop, V>(table: &SmartTable<K, V>, key: K): bool {
```

Comment:
What I'm missing here is a good way to do a lookup if it's found or return false if not. A lot of the time you'd like to do things like "lookup if key exists otherwise insert a new value". With this API you seem to be looking up keys spuriously. I think something like a signature of

Pseudo-code:
Lookup(table: &SmartTable, key: K): Optional<&V> 

would be most useful

Reply:
I don't disagree with you. But here we're trying to make API consistent with `Table` so ppl can use them interchangeably. Otherwise I would do the same as you suggested.